### PR TITLE
Resolve dealloc problem and multi thread problem in PinnedMemory

### PR DESCRIPTION
**Pull request by okuta**
Please see #381 and #382.

### CODE REVIEW

File: cupy/cuda/pinned_memory.pyx [hunk 1]
```diff
@@ -271,42 +276,49 @@ cdef class PinnedMemoryPool:
...
+            mem = self._in_use.pop(ptr, None)
+            if mem is None:
+                raise RuntimeError('Cannot free out-of-pool memory')
+            free = self._free[size]
+            free.append(mem)
+        finally:
+            rlock.unlock_fastrlock(self._lock)
 
     cpdef free_all_blocks(self):
         """Release free all blocks."""
```
Comment (sonots) [inline cupy/cuda/pinned_memory.pyx]: This is a public method, so do we need locking inside this, too?

Comment (okuta) [reply cupy/cuda/pinned_memory.pyx]: I fixed.

**Comment by sonots** [comment]
jenkins, test this please

**Comment by sonots** [comment]
jenkins, test this please

**Pull request merged**